tools/wptrunner/wptrunner/testrunner.py

```diff
@@ -1020,18 +1038,32 @@ def wait(self, timeout: Optional[float] = None) -> None:
             timeout: Overall timeout (in seconds) for all threads to join. The
                 default value indicates an indefinite timeout.
         """
-        deadline = None if timeout is None else time.time() + timeout
-        for manager in self.pool:
-            manager_timeout = None
-            if deadline is not None:
-                manager_timeout = max(0, deadline - time.time())
-            manager.join(manager_timeout)
+        # Here, the main thread cannot simply `join()` the threads in
+        # `self.pool` sequentially because a keyboard interrupt raised during a
+        # `Thread.join()` may incorrectly mark that thread as "stopped" when it
+        # is not [0, 1]. Subsequent `join()`s for the affected thread won't
+        # block anymore, so a subsequent `ManagerGroup.wait()` may return with
+        # that thread still alive.
+        #
+        # To the extent the timeout allows, it's important that
+        # `ManagerGroup.wait()` returns with all `TestRunnerManager` threads
+        # actually stopped. Otherwise, a live thread may log after `mozlog`
+        # shutdown (not allowed) or worse, leak browser processes that the
+        # thread should have stopped when exiting its run loop [2].
+        #
+        # [0]: https://github.com/python/cpython/issues/90882
+        # [1]: https://github.com/python/cpython/blob/558b517b/Lib/threading.py#L1146-L1178
+        # [2]: https://crbug.com/330236796
+        assert self.stop_state, "ManagerGroup hasn't been started yet"
+        assert not self.stop_state.done.broken, self.stop_state
+        self.stop_state.done.wait(timeout)
 
     def stop(self):
         """Set the stop flag so that all managers in the group stop as soon
         as possible"""
-        self.stop_flag.set()
-        self.logger.debug("Stop flag set in ManagerGroup")
+        if self.stop_state:
+            self.stop_state.should_stop.set()
```

Comment:
How about we call this "stop_flag", and implement the various functions for it? We could then use it similarly as the previous flag.